csrc/quantization/cutlass_w8a8/grouped_mm_c3x.cu

```diff
@@ -0,0 +1,331 @@
+                                  const int topk) {
+  int expert_id = blockIdx.x;
+
+  for (int i = threadIdx.x; i < topk_length; i += THREADS_PER_EXPERT) {
+    if (topk_ids[i] == expert_id) {
+      int start = atomicAdd(&atomic_buffer[expert_id], 1);
+      arg_sort[start] = i / topk;
+      arg_sort_prim[i] = start;
+    }
+  }
+}
+
+constexpr int THREADS_PER_EXPERT_2 = 32;
+
+// 1 warp per expert
+// 4 experts per block
+__global__ void compute_problem_sizes_multi_expert(
+    const int* __restrict__ topk_ids, int32_t* problem_sizes1,
+    int32_t* problem_sizes2, int32_t* atomic_buffer, const int topk_length,
+    const int n, const int k) {
+  int expert_id = blockIdx.x * 4 + threadIdx.x / THREADS_PER_EXPERT_2;
+  int start = threadIdx.x % THREADS_PER_EXPERT_2;
+
+  int occurrences = 0;
+  for (int i = start; i < topk_length; i += THREADS_PER_EXPERT_2) {
+    occurrences += (topk_ids[i] == expert_id);
+  }
+  atomicAdd(&atomic_buffer[expert_id], occurrences);
+  // we only need this if #threads/expert > warp_size
+  if constexpr (THREADS_PER_EXPERT_2 > 32) {
+    __syncthreads();
+  }
+
+  if (start == 0) {
+    int final_occurrences = atomic_buffer[expert_id];
+    problem_sizes1[expert_id * 3] = final_occurrences;
+    problem_sizes1[expert_id * 3 + 1] = 2 * n;
+    problem_sizes1[expert_id * 3 + 2] = k;
+    problem_sizes2[expert_id * 3] = final_occurrences;
+    problem_sizes2[expert_id * 3 + 1] = k;
+    problem_sizes2[expert_id * 3 + 2] = n;
+  }
+}
+
+__global__ void compute_arg_sorts_multi_expert(
+    const int* __restrict__ topk_ids, int32_t* arg_sort, int32_t* arg_sort_prim,
+    int32_t* atomic_buffer, const int topk_length, const int topk) {
+  int expert_id = blockIdx.x * 4 + threadIdx.x / THREADS_PER_EXPERT_2;
+  int start = threadIdx.x % THREADS_PER_EXPERT_2;
+
+  for (int i = start; i < topk_length; i += THREADS_PER_EXPERT_2) {
+    if (topk_ids[i] == expert_id) {
+      int start = atomicAdd(&atomic_buffer[expert_id], 1);
+      arg_sort[start] = i / topk;
+      arg_sort_prim[i] = start;
+    }
+  }
+}
+
+void get_grouped_mm_data_caller(
+    const torch::Tensor& topk_ids, torch::Tensor& expert_offsets,
+    torch::Tensor& problem_sizes1, torch::Tensor& problem_sizes2,
+    torch::Tensor& arg_sort, torch::Tensor& arg_sort_prim,
+    const int64_t num_experts, const int64_t n, const int64_t k) {
+  auto stream = at::cuda::getCurrentCUDAStream(topk_ids.device().index());
+  auto options_int32 =
+      torch::TensorOptions().dtype(torch::kInt32).device(topk_ids.device());
+  torch::Tensor atomic_buffer = torch::zeros(num_experts, options_int32);
+
+  // TODO this is an alternative way to block kernels
+  constexpr bool multi_expert_blocks = false;
+  if constexpr (multi_expert_blocks) {
+    int num_blocks = (num_experts + 3) / 4;
+    int num_threads = THREADS_PER_EXPERT_2 * 4;
+    compute_problem_sizes_multi_expert<<<num_blocks, num_threads, 0, stream>>>(
+        (const int32_t*)topk_ids.data_ptr(),
```

Comment:
Could these casts be an argument to `data_ptr` instead of C-style casts? Like `.template data_ptr<const int32_t>`